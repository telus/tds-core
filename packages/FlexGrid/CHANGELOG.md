# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [3.0.11](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.10...@tds/core-flex-grid@3.0.11) (2020-07-22)

**Note:** Version bump only for package @tds/core-flex-grid





## [3.0.10](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.9...@tds/core-flex-grid@3.0.10) (2020-02-24)

**Note:** Version bump only for package @tds/core-flex-grid





## [3.0.9](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.8...@tds/core-flex-grid@3.0.9) (2020-02-24)


### Bug Fixes

* **core-flex-grid:** prevent invisible hairline divider w/ mdOffset+ ([9cdec38](https://github.com/telusdigital/tds/commit/9cdec38))





## [3.0.8](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.7...@tds/core-flex-grid@3.0.8) (2020-02-06)


### Bug Fixes

* **core-flex-grid:** prevent incorrect width in box ([653a725](https://github.com/telusdigital/tds/commit/653a725))





## [3.0.7](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.6...@tds/core-flex-grid@3.0.7) (2020-02-03)

**Note:** Version bump only for package @tds/core-flex-grid





## [3.0.6](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.5...@tds/core-flex-grid@3.0.6) (2020-01-27)

**Note:** Version bump only for package @tds/core-flex-grid





## [3.0.5](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.4...@tds/core-flex-grid@3.0.5) (2020-01-23)

**Note:** Version bump only for package @tds/core-flex-grid





## [3.0.4](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.3...@tds/core-flex-grid@3.0.4) (2020-01-22)

**Note:** Version bump only for package @tds/core-flex-grid





## [3.0.3](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.2...@tds/core-flex-grid@3.0.3) (2019-09-06)

**Note:** Version bump only for package @tds/core-flex-grid





## [3.0.2](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.1...@tds/core-flex-grid@3.0.2) (2019-08-22)

**Note:** Version bump only for package @tds/core-flex-grid





## [3.0.1](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@3.0.0...@tds/core-flex-grid@3.0.1) (2019-08-21)


### Bug Fixes

* **core-flex-grid:** convert px units to rem in flexgrid container ([0da7ffc](https://github.com/telusdigital/tds/commit/0da7ffc))





# [3.0.0](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.4.1...@tds/core-flex-grid@3.0.0) (2019-07-26)


### Features

* **core-flex-grid:** upgrade to styled-components ([8bf525a](https://github.com/telusdigital/tds/commit/8bf525a))


### BREAKING CHANGES

* **core-flex-grid:** add styled-components as a peerDependency





## [2.4.1](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.4.0...@tds/core-flex-grid@2.4.1) (2019-07-26)

**Note:** Version bump only for package @tds/core-flex-grid





# [2.4.0](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.13...@tds/core-flex-grid@2.4.0) (2019-07-25)


### Features

* **core-flex-grid:** upgrade to styled-components ([734f951](https://github.com/telusdigital/tds/commit/734f951))





## [2.3.13](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.12...@tds/core-flex-grid@2.3.13) (2019-06-04)

**Note:** Version bump only for package @tds/core-flex-grid

## [2.3.12](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.11...@tds/core-flex-grid@2.3.12) (2019-05-13)

**Note:** Version bump only for package @tds/core-flex-grid

## [2.3.11](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.10...@tds/core-flex-grid@2.3.11) (2019-05-01)

**Note:** Version bump only for package @tds/core-flex-grid

## [2.3.10](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.9...@tds/core-flex-grid@2.3.10) (2019-04-22)

**Note:** Version bump only for package @tds/core-flex-grid

## [2.3.9](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.8...@tds/core-flex-grid@2.3.9) (2019-04-04)

**Note:** Version bump only for package @tds/core-flex-grid

## [2.3.8](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.7...@tds/core-flex-grid@2.3.8) (2019-03-28)

**Note:** Version bump only for package @tds/core-flex-grid

## [2.3.7](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.6...@tds/core-flex-grid@2.3.7) (2019-03-13)

### Bug Fixes

- **core-flex-grid:** downgrade react-broadcast ([70864e4](https://github.com/telusdigital/tds/commit/70864e4))

## [2.3.6](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.5...@tds/core-flex-grid@2.3.6) (2019-03-11)

**Note:** Version bump only for package @tds/core-flex-grid

## [2.3.5](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.4...@tds/core-flex-grid@2.3.5) (2019-02-07)

### Bug Fixes

- **core-flex-grid:** adjust css output to be mobile-first ([431cb43](https://github.com/telusdigital/tds/commit/431cb43)), closes [#899](https://github.com/telusdigital/tds/issues/899)

## [2.3.4](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.3...@tds/core-flex-grid@2.3.4) (2019-02-06)

**Note:** Version bump only for package @tds/core-flex-grid

## [2.3.3](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.2...@tds/core-flex-grid@2.3.3) (2019-01-10)

**Note:** Version bump only for package @tds/core-flex-grid

<a name="2.3.2"></a>

## [2.3.2](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.1...@tds/core-flex-grid@2.3.2) (2018-12-13)

### Bug Fixes

- **core-flex-grid:** apply flex-shrink 0 to FlexGrid.Row ([326c99b](https://github.com/telusdigital/tds/commit/326c99b)), closes [#783](https://github.com/telusdigital/tds/issues/783)

<a name="2.3.1"></a>

## [2.3.1](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.3.0...@tds/core-flex-grid@2.3.1) (2018-12-11)

**Note:** Version bump only for package @tds/core-flex-grid

<a name="2.3.0"></a>

# [2.3.0](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.2.1...@tds/core-flex-grid@2.3.0) (2018-12-10)

### Features

- **core-flex-grid:** add horizontalAlign prop to Col ([694ad0a](https://github.com/telusdigital/tds/commit/694ad0a))

<a name="2.2.1"></a>

## [2.2.1](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.2.0...@tds/core-flex-grid@2.2.1) (2018-12-06)

**Note:** Version bump only for package @tds/core-flex-grid

<a name="2.2.0"></a>

# [2.2.0](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.1.1...@tds/core-flex-grid@2.2.0) (2018-08-09)

### Features

- **core-flex-grid:** add hide props to col ([1f4ed86](https://github.com/telusdigital/tds/commit/1f4ed86))
- **core-flex-grid:** hides column starting from breakpoint ([7d6f267](https://github.com/telusdigital/tds/commit/7d6f267))
- **core-flex-grid:** more abstraction in calculateLevels ([faec59d](https://github.com/telusdigital/tds/commit/faec59d))
- **core-flex-grid:** removed hidden flags and enabled 0 for breakpoints ([867599e](https://github.com/telusdigital/tds/commit/867599e))

<a name="2.1.1"></a>

## [2.1.1](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.1.0...@tds/core-flex-grid@2.1.1) (2018-08-07)

### Bug Fixes

- **core-flex-grid:** add specificity to styles ([ee0d7ff](https://github.com/telusdigital/tds/commit/ee0d7ff))

<a name="2.1.0"></a>

# [2.1.0](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.0.2...@tds/core-flex-grid@2.1.0) (2018-07-17)

### Features

- **core-flex-grid:** add reversible grid props ([#681](https://github.com/telusdigital/tds/issues/681)) ([9d10ec6](https://github.com/telusdigital/tds/commit/9d10ec6))

<a name="2.0.2"></a>

## [2.0.2](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.0.1...@tds/core-flex-grid@2.0.2) (2018-06-12)

**Note:** Version bump only for package @tds/core-flex-grid

<a name="2.0.1"></a>

## [2.0.1](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@2.0.0...@tds/core-flex-grid@2.0.1) (2018-06-12)

### Bug Fixes

- **core-flex-grid:** force correct spacing through specificity ([da485d4](https://github.com/telusdigital/tds/commit/da485d4))

<a name="2.0.0"></a>

# [2.0.0](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@1.2.1...@tds/core-flex-grid@2.0.0) (2018-06-04)

### Features

- **core-flex-grid:** make limitWidth true by default ([9ce0b69](https://github.com/telusdigital/tds/commit/9ce0b69))
- **core-flex-grid:** remove centre prop ([45ebff9](https://github.com/telusdigital/tds/commit/45ebff9))

### BREAKING CHANGES

- **core-flex-grid:** remove the centre prop
- **core-flex-grid:** set limitWidth prop to true by default

<a name="1.2.1"></a>

## [1.2.1](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@1.2.0...@tds/core-flex-grid@1.2.1) (2018-05-14)

### Bug Fixes

- **core-flex-grid:** remove old responsive sass ([e21a59c](https://github.com/telusdigital/tds/commit/e21a59c))

<a name="1.2.0"></a>

# [1.2.0](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@1.1.0...@tds/core-flex-grid@1.2.0) (2018-04-11)

### Features

- **core-flex-grid:** add responsive props ([#566](https://github.com/telusdigital/tds/issues/566)) ([10e492c](https://github.com/telusdigital/tds/commit/10e492c)), closes [#465](https://github.com/telusdigital/tds/issues/465)

<a name="1.1.0"></a>

# [1.1.0](https://github.com/telusdigital/tds/compare/@tds/core-flex-grid@1.0.0...@tds/core-flex-grid@1.1.0) (2018-03-22)

### Features

- **core-flex-grid:** fix gutterless alignment issue and deprecate centre prop ([#526](https://github.com/telusdigital/tds/issues/526)) ([0e8cbeb](https://github.com/telusdigital/tds/commit/0e8cbeb)), closes [#518](https://github.com/telusdigital/tds/issues/518)
