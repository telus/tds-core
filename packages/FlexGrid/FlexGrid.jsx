import React from 'react'
import PropTypes from 'prop-types'
import { Broadcast } from 'react-broadcast'
import { Grid } from 'react-flexbox-grid'

import Col from './Col/Col'
import Row from './Row/Row'

import safeRest from '../../shared/utils/safeRest'
import joinClassNames from '../../shared/utils/joinClassNames'

import styles from './FlexGrid.modules.scss'

/**
 * A mobile-first flexbox grid.
 *
 * @version ./package.json
 */
const FlexGrid = ({ limitWidth, gutter, children, ...rest }) => {
  return (
    <Broadcast channel="flex-grid" value={gutter}>
      <Grid
        {...safeRest(rest)}
        fluid
        className={joinClassNames(styles.flexGrid, limitWidth && styles.limitWidth)}
      >
        {children}
      </Grid>
    </Broadcast>
  )
}

FlexGrid.propTypes = {
  /**
   * Whether or not to give the grid a fixed width. This also centres the grid horizontally.
   */
  limitWidth: PropTypes.bool,
  /**
   * Whether or not to include gutters in between columns.
   */
  gutter: PropTypes.bool,
  /**
   * The rows and columns of the Grid. Will typically be `FlexGrid.Row` and `FlexGrid.Col` components.
   */
  children: PropTypes.node.isRequired,
}

FlexGrid.defaultProps = {
  limitWidth: true,
  gutter: true,
}

FlexGrid.Row = Row
FlexGrid.Col = Col

export default FlexGrid
