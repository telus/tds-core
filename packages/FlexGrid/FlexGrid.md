**Use the "Open isolated" button above to view this component in full-width mode.**

The `FlexGrid` system is a thin wrapper over [react-flexbox-grid](https://github.com/roylee0704/react-flexbox-grid),
which implements [flexboxgrid.css](http://flexboxgrid.com/). It uses a series of rows and columns to layout and align
content 2-dimensionally.

**If you are new to or unfamiliar with flexbox, read this [CSS Tricks flexbox guide](https://css-tricks.com/snippets/css/a-guide-to-flexbox/)
for background, terminology, guidelines, and code snippets.**

Below is a quick example of how the components fit together. See the [FlexGrid.Row](#row) and [FlexGrid.Col](#col)
components below for a more in-depth look.

```jsx { "props": { "className": "docs_full-width-playground docs_flex-grid-coloring" } }
<FlexGrid>
  <FlexGrid.Row>
    <FlexGrid.Col>
      <Box vertical={2}><Text>1/3</Text></Box>
    </FlexGrid.Col>
    <FlexGrid.Col>
      <Box vertical={2}><Text>2/3</Text></Box>
    </FlexGrid.Col>
    <FlexGrid.Col>
      <Box vertical={2}><Text>3/3</Text></Box>
    </FlexGrid.Col>
  </FlexGrid.Row>
</FlexGrid>
```

### How it works

* `FlexGrid` is the top-level container. It centers and horizontally pads your siteâ€™s contents
* `FlexGrid.Row`s are wrappers for columns. Columns have horizontal padding (called a gutter) for controlling the space between them
* Place content inside `FlexGrid.Col` components. These should be the immediate children of rows
* Columns without a specified width will automatically layout as equal width columns
* Column widths can be made responsive by using the responsive props, which match the props used in `Responsive`: `xs`, `sm`,
  `md`, `lg` and `xl`
* The responsive column widths are based on minimum width media queries, so they apply to the specified breakpoint and all
  those above it

### Removing the gutter

Remove the 16px gutter surrounding columns by passing `gutter={false}` to the `FlexGrid` container. This removes the horizontal padding
from all children columns.

```jsx { "props": { "className": "docs_full-width-playground docs_flex-grid-coloring" } }
<FlexGrid gutter={false}>
  <FlexGrid.Row>
    <FlexGrid.Col>
      <Box vertical={2}><Text>1/3 no gutter</Text></Box>
    </FlexGrid.Col>
    <FlexGrid.Col>
      <Box vertical={2}><Text>2/3 no gutter</Text></Box>
    </FlexGrid.Col>
    <FlexGrid.Col>
      <Box vertical={2}><Text>3/3 no gutter</Text></Box>
    </FlexGrid.Col>
  </FlexGrid.Row>
</FlexGrid>
```
