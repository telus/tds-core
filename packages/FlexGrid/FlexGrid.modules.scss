@import '~@tds/core-responsive/responsive';

.flexGrid {
  padding: 0;
}

.limitWidth {
  margin: 0 auto;
  @each $viewport, $mq-breakpoint in $mq-breakpoints {
    @include mq($from: $viewport) {
      max-width: $mq-breakpoint;
    }
  }
}
