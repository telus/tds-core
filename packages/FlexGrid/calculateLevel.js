const calculateLevel = (xs, sm, md, lg, xl) => {
  const levelToggles = [xs, sm, md, lg, xl]
  const enabledLevels = [false, false, false, false, false]

  for (let toggles = 0; toggles < levelToggles.length; toggles += 1) {
    for (let levels = toggles; levels < enabledLevels.length; levels += 1) {
      if (levelToggles[toggles] !== undefined) {
        enabledLevels[levels] = levelToggles[toggles]
      }
    }
  }

  return enabledLevels
}

export default calculateLevel
