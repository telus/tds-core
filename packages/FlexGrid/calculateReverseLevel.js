const calculateReverseLevel = (xsReverse, smReverse, mdReverse, lgReverse, xlReverse) => {
  const levelToggles = [xsReverse, smReverse, mdReverse, lgReverse, xlReverse]
  const enabledLevels = [false, false, false, false, false]

  for (let toggles = 0; toggles < levelToggles.length; toggles += 1) {
    for (let levels = toggles; levels < enabledLevels.length; levels += 1) {
      if (levelToggles[toggles] !== undefined) {
        enabledLevels[levels] = levelToggles[toggles]
      }
    }
  }

  return enabledLevels
}

export default calculateReverseLevel
