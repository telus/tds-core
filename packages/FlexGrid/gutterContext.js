import React from 'react'

const GutterContext = React.createContext(false)

export default GutterContext
