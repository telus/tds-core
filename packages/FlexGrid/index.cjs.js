const FlexGrid = require('./dist/index.cjs')

module.exports = FlexGrid
