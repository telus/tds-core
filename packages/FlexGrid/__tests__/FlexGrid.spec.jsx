import React from 'react'
import { mount, render } from 'enzyme'
import { Grid } from 'react-flexbox-grid'

import FlexGrid from '../FlexGrid'

import mockMatchMedia from '../../../config/jest/__mocks__/matchMedia'

describe('FlexGrid', () => {
  const doMount = (props = {}) => {
    const flexGrid = mount(
      <FlexGrid {...props}>
        <FlexGrid.Row id="row-1">
          <FlexGrid.Col id="col-1">1st column content</FlexGrid.Col>
          <FlexGrid.Col id="col-2">2nd column content</FlexGrid.Col>
        </FlexGrid.Row>
      </FlexGrid>
    )

    return {
      flexGrid: flexGrid.find(Grid),
      findColumn: index => flexGrid.find(`div#col-${index}`),
      findRow: index => flexGrid.find(`div#row-${index}`),
    }
  }

  it('renders', () => {
    const flexGrid = render(
      <FlexGrid>
        <FlexGrid.Row>
          <FlexGrid.Col>1st column content</FlexGrid.Col>
          <FlexGrid.Col>2nd column content</FlexGrid.Col>
        </FlexGrid.Row>
      </FlexGrid>
    )

    expect(flexGrid).toMatchSnapshot()
  })

  it('renders a react-flexbox-grid Grid component', () => {
    const { flexGrid } = doMount()

    expect(flexGrid).toMatchSelector(Grid)
  })

  it('is fluid by default', () => {
    const { flexGrid } = doMount()

    expect(flexGrid).toHaveProp('fluid', true)
  })

  it('should render all children related to Grid with no gutter', () => {
    const { findColumn, findRow } = doMount({ gutter: false })

    const col1 = findColumn(1)
    const col2 = findColumn(2)
    const row = findRow(1)

    expect(col1).toHaveClassName('gutterless')
    expect(col2).toHaveClassName('gutterless')
    expect(row).toHaveClassName('flexRow')
  })

  it('passes additional attributes to the element', () => {
    const { flexGrid } = doMount({
      id: 'the-id',
      'data-some-attr': 'some value',
    })

    expect(flexGrid).toHaveProp('id', 'the-id')
    expect(flexGrid).toHaveProp('data-some-attr', 'some value')
  })

  it('can have limited width at various viewport sizes', () => {
    mockMatchMedia(360)
    const { flexGrid } = doMount()

    expect(flexGrid).toHaveClassName('limitWidth')
  })

  it('does not allow custom CSS', () => {
    const { flexGrid } = doMount({
      className: 'my-custom-class',
      style: { color: 'hotpink' },
    })

    expect(flexGrid).not.toHaveProp('className', 'my-custom-class')
    expect(flexGrid).not.toHaveProp('style', { color: 'hotpink' })
  })

  it('allows the limitWidth prop to be unset', () => {
    const { flexGrid } = doMount({
      limitWidth: false,
    })

    expect(flexGrid).not.toHaveClassName('limitWidth')
  })
})
