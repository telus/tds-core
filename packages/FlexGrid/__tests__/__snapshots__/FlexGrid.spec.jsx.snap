// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlexGrid renders 1`] = `
<div
  class="flexGrid limitWidth container-fluid"
>
  <div
    class="flexRow row"
  >
    <div
      class="col-xs padding"
    >
      1st column content
    </div>
    <div
      class="col-xs padding"
    >
      2nd column content
    </div>
  </div>
</div>
`;
